### DISCUSSION TITLE
Improvement request for better granulation of exceptions

### DISCUSSION THREAD

**Discussion by anderso790** [OxalisCommunity/Oxalis-AS4#158] [Ideas]
We would like improvements in the granularity of exceptions in oxalis code in general, but especially in Oxalis-AS4. 

It would be very helpful if we at first could get categorization of exceptions that: 
1- Delivery failed but resending possible: Exceptions that means that resending is possible
2- Delivery failed permanently: Exceptions that means that resending will not help. 

In general better exceptions will be very helpful in the day to day operation of oxalis AP, but I guess this will also help in understanding/analyzing issues reported in github.

**Comment by aaron-kumar** [comment]
Thanks for raising this issue @anderso790 . 
Yes, we have plan to categorize the type of exception.

Below is my input (_not extensive yet, but good to start with_):

- Communication error: Sending Access Point won't be able to deliver document to Receiving Access Point because of transport/communication problem (e.g. Receiving AP goes down while delivering or SMP issue). -- Recoverable error and can be document can be retried/resend. 
- Basic structure Issue: Document can't be delivered to receiving access point because there is problem with document itself OR in the registered SMP document and/or transport profiles. Here Oxalis can't deliver document document . -- Unrecoverable error , action must be required before resending can be tried. 
- Configuration Issue: In case, Oxalis is not configured correctly  -- Unrecoverable error, action must be required before resending is possible
- Unknown Issue: Additional issues which can't be categorized due to lack of information.

We need more input from Oxalis community member here, so I am converting this "issue" to "discussion" unless we come up with concrete plan

**Reply by aaron-kumar** [reply]
**Request to Oxalis Community Member**: Please vote (thumbs up) for this issue if you think this will solve your lots of problems.

**Reply by teedjay** [reply]
Good suggestions @aaron-kumar!  Would be great if we could look at a category to see if :   
- failed, message cannot be delivered at all (no need to retry)
- failed, message can be delivered later (some transient issues right now)
